iio:adxl372: New driver for Analog Devices ADXL372 Accelerometer

stefpopa opened the pull request:
This patch adds basic support for Analog Devices ADXL372 SPI-Bus Three-Axis
Digital Accelerometer.

The device is probed and configured the with some initial default values.
With this basic driver, it is possible to read raw acceleration data.

Datasheet:
http://www.analog.com/media/en/technical-documentation/data-sheets/ADXL372.pdf

Signed-off-by: Stefan Popa <stefan.popa@analog.com>

stefpopa commented:
This is a basic version of the ADXL372 driver. Next versions will add more features such as support for 
- FIFO
- IRQ
- ODR
 - BW

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -0,0 +1,624 @@
+
+	switch (info) {
+	case IIO_CHAN_INFO_RAW:
+		mutex_lock(&indio_dev->mlock);
```
i think it may be incorrect to use this lock here ; since this may be a lock that should not be managed by the driver, but rather the IIO infrastructure ; 
i could be wrong ;
@larsclausen  ?

stefpopa replied:
I think that you are right, regmap_raw_read already does this: 
https://elixir.bootlin.com/linux/v3.1/source/drivers/base/regmap/regmap.c#L374

lclausen-adi replied:
There is currently an effort to make this lock subsystem only. You'll still find some old drivers that use it, but new drivers should not use it.

commodo replied:
FWIW: the `mlock` thing here may be a broader discussion regarding IIO infrastructure ;and there are other drivers that seem to use `mlock` in similar sections like this one

lclausen-adi replied:
@commodo There are some people working on updating drivers to not use `mlock` anymore. If you want to join than have fun with it :)

commodo replied:
ah ; will see ; maybe after some of my learning curve has attenuated

dbogdan commented on drivers/iio/accel/adxl372.c:
```diff
@@ -0,0 +1,624 @@
+#include <linux/interrupt.h>
+#include <linux/irq.h>
+#include <linux/acpi.h>
+#include <linux/gpio/consumer.h>
```
Some header files are unnecessary to be included - e.g. gpio/consumer.h, irq.h.

lclausen-adi commented on drivers/iio/accel/adxl372.c:
```diff
@@ -0,0 +1,624 @@
+#include <linux/iio/triggered_buffer.h>
+#include <linux/iio/trigger_consumer.h>
+
+#define ADXL372_DRV_NAME		"adxl372"
```
We don't really need this macro, it is only used once.

lclausen-adi commented on arch/arm/boot/dts/overlays/rpi-adxl372-overlay.dts:
```diff
@@ -0,0 +1,21 @@
+/dts-v1/;
```
This file should be in a separate commit.

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -0,0 +1,624 @@
+
+	/* scale factor is 100 mg/code */
+	th_reg_high_val = (threshold / 100) >> 3;
+	th_reg_low_val = ((threshold / 100) << 5) | (referenced << 1) | enable;
```
i'm a bit unsure about masking and shifting `bool` types;
it may be fine ; but it's not very common ;

maybe just use `u8` ; or define constants that you pass here;
maybe something like
```
#define ADXL372_THRESHOLD_EN		                        BIT(0)
#define ADXL372_THRESHOLD_REFERENCED_EN		BIT(1)
```

and call something like
```
adxl372_set_activity_threshold(....,  ADXL372_THRESHOLD_REFERENCED_EN,  ADXL372_THRESHOLD_EN) 
```

lclausen-adi commented on drivers/iio/accel/Kconfig:
```diff
@@ -5,6 +5,18 @@
 
+config ADXL372
+    tristate "Analog Devices ADXL372 3-Axis Accelerometer Driver"
+	depends on SPI
```
Indentation looks strange here

stefpopa commented:
changelog v1->v2:
- removed mlocks
- removed unnecessary includes
- modified the adxl372_set_activity_threshold function, such that shifting bool types will be avoided
- removed unnecessary ADXL372_DRV_NAME macro
- created a separate commit for the rpi-adxl372-overlay.dts file
- fixed the indentation for Kconfig file
- created a #define for the scale instead of using a const int
